csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -100,23 +152,38 @@ public sealed class OutgoingRequestFrame : OutputStream
                                     IReadOnlyDictionary<string, string>? context, ArraySegment<byte> payload)
             : this(proxy, operation, idempotent, context)
         {
-            if (payload.Count == 0)
-            {
-                WriteEmptyEncapsulation(_payloadEncoding);
-            }
-            else
+            var ostr = new OutputStream(Encoding, Data, _payloadStart);
+            _payloadEnd = ostr.WriteEncapsulation(payload);
+            Size = Data.GetByteCount();
+            IsSealed = true;
+        }
+
+        /// <summary>Creates and returns an OutputStream that can be used to write the payload of this request,
+        /// once the caller finish writing the payload it must call SavePayload passing the returned
+        /// stream as argument.</summary>
+        /// <param name="format">The format type for the payload.</param>
+        /// <returns>An OutputStream instance that can be used to write the payload of this request.</returns>
+        public OutputStream WritePayload(FormatType? format = null)
+        {
+            if (_payloadEnd != null)
             {
-                WritePayload(payload);
+                throw new InvalidOperationException("the frame already contains a payload");
             }
+
+            return new OutputStream(Encoding, Data, _payloadStart, true, format,
+                paylaodEnd => PayloadReady(this, paylaodEnd));
```

Comment:
The idea was to not capture this to allow reusing the same lambda, we are still capturing it so we can use a instance method, or pass an extra argument to the lambda, it can be an `object frame` given that we don't have a common base class.